qiskit/synthesis/unitary/qsd.py

```diff
@@ -120,52 +146,125 @@ def decomp_2q(mat):
+                    decirc, _, _ = _demultiplex(
                         um00,
                         um11,
                         opt_a1=opt_a1,
                         opt_a2=opt_a2,
+                        _vw_type="all",
                         _depth=_depth,
                         _ctrl_index=nqubits - 1 - ctrl_index,
                     )
                     return decirc
         qr = QuantumRegister(nqubits)
-        circ = QuantumCircuit(qr)
-        # perform cosine-sine decomposition
-        (u1, u2), vtheta, (v1h, v2h) = cossin(np.asarray(mat, dtype=complex))
+
+        # perform block ZXZ decomposition from [2]
+        A1, A2, B, C = _block_zxz_decomp(np.asarray(mat, dtype=complex))
+        iden = np.eye(2 ** (nqubits - 1))
         # left circ
-        left_circ = _demultiplex(v1h, v2h, opt_a1=opt_a1, opt_a2=opt_a2, _depth=_depth)
-        circ.append(left_circ.to_instruction(), qr)
+        left_circ, vmatC, wmatC = _demultiplex(
+            iden, C, opt_a1=opt_a1, opt_a2=opt_a2, _vw_type="only_w", _depth=_depth
+        )
+        # right circ
+        right_circ, vmatA, wmatA = _demultiplex(
+            A1, A2, opt_a1=opt_a1, opt_a2=opt_a2, _vw_type="only_v", _depth=_depth
+        )
+
         # middle circ
+        # zmat is needed in order to reduce two cz gates, and combine them into the B2 matrix
+        zmat = np.diag([1] * (dim // 4) + [-1] * (dim // 4))
+        # wmatA and vmatC are combined into B1 and B2
+        B1 = wmatA @ vmatC
         if opt_a1:
-            nangles = len(vtheta)
-            half_size = nangles // 2
-            # get UCG in terms of CZ
-            circ_cz = _get_ucry_cz(nqubits, (2 * vtheta).tolist())
-            circ.append(circ_cz.to_instruction(), range(nqubits))
-            # merge final cz with right-side generic multiplexer
-            u2[:, half_size:] = np.negative(u2[:, half_size:])
+            B2 = zmat @ wmatA @ B @ vmatC @ zmat
         else:
-            ucry = UCRYGate((2 * vtheta).tolist())
-            circ.append(ucry, [qr[-1]] + qr[:-1])
-        # right circ
-        right_circ = _demultiplex(u1, u2, opt_a1=opt_a1, opt_a2=opt_a2, _depth=_depth)
+            B2 = wmatA @ B @ vmatC
+        middle_circ, _, _ = _demultiplex(
+            B1, B2, opt_a1=opt_a1, opt_a2=opt_a2, _vw_type="all", _depth=_depth
+        )
+
+        # the output circuit of the block ZXZ decomposition from [2]
+        circ = QuantumCircuit(qr)
+        circ.append(left_circ.to_instruction(), qr)
+        circ.h(nqubits - 1)
+        circ.append(middle_circ.to_instruction(), qr)
+        circ.h(nqubits - 1)
         circ.append(right_circ.to_instruction(), qr)
 
     if opt_a2 and _depth == 0 and dim > 4:
         return _apply_a2(circ)
     return circ
 
 
-def _demultiplex(um0, um1, opt_a1=False, opt_a2=False, *, _depth=0, _ctrl_index=None):
+def _block_zxz_decomp(Umat):
+    """Block ZXZ decomposition method, by Krol and Al-Ars [2]."""
+    N = Umat.shape[0]
+    n = N // 2
```

Comment:
Nit: I'd find something like this a bit easier to read. The difference between `N` and `n` was a bit confusing especially since you're explicitly saying this isn't a notation in the paper (as the comment below says the notation starts after this).

```suggestion
    dim = Umat.shape[0]
    n = dim // 2
```